### ISSUE TITLE
Assets not loading correctly for installations in subfolders

### ISSUE THREAD

**Issue by prbt2016** [bolt/bolt#6071]
Hi, 
Prashant here . I tested Bolt 3.2.2 on centos linux 6.4 on php version 5.5 . The frontend theme is not loading properly.What could be the issue .I am uploading a screenshot of the same.I tried configuring .htaccess too as written in the documentation.Still i get the same error.This is occuring since version 3.2.0 .Before this version it didn't require to configure .htaccess .The backend works and loads  fine .I opened the debug panel in firebug and it shows that it's not redirecting properly to the theme folder.Could you please try to replicate the issue at your side.Waiting for your kind reply
![bolt_theme](https://cloud.githubusercontent.com/assets/17871448/20550555/e20f02ca-b15c-11e6-8052-5891b4231229.png)
![debug](https://cloud.githubusercontent.com/assets/17871448/20550558/f222d6d2-b15c-11e6-92d4-f573f7fb6d1c.png)

**Comment by bobdenotter** [comment]
How did you install Bolt?

**Comment by prbt2016** [comment]
Installed it manually in subfolder.Created the database first(named bolt) and then  configured the config.yml file (in /app/config) and entered db details as follows  in it (configured it with mysql)
  driver: mysql
  username: root
  password: mypass
  databasename: bolt
  host: localhost
then hit the url on browser http://domain/subfolder/bolt/public and entered the user details  .

**Comment by bobdenotter** [comment]
Did you clone it from git? Use a composer install? Download the .tar.gz or .zip from somewhere? 

Looks like you're using subfolders, instead of putting it in the "root", so you might be seeing the same as this person: #5924

**Comment by prbt2016** [comment]
downloaded the zip version from here https://bolt.cm/pages/download ......

**Comment by GawainLynch** [comment]
Either your document root is set incorrectly in your web server configuration, or you're using the local PHP development server, would be my two guesses.

**Comment by prbt2016** [comment]
Hi,
I tested it on hosting (cpanel) on php 5.5.got the same issue.there too I installed it in a subdirectory.I never had this issue for versions of bolt below 3.2.0.The back end is displayed properly.does the .htaccess in the public folder need special configurations ?....I tried changing RewriteBase in that to relativeurl I.e /subfolder/boltdirname ....still didn't work .

**Comment by trollix** [comment]
Hello,
Same bug for me, in 3.2.2 and 3.2.3 versions.
Install by sftp in a subdirectory /bolt/boltv3/ 
It's working with  other themes, for example, with bolt-theme-solarize .
Hope it's helps,
Trollix

**Comment by HorlogeSkynet** [comment]
Same bug for me !
Installed with the .tar.gz...

**Comment by bobdenotter** [comment]
> It's working with other themes, for example, with bolt-theme-solarize

Am I understanding correctly that _some_ themes work correctly, while _others_ don't? If so, that might be a problem with the theme itself. 

What theme doesn't work?

**Comment by prbt2016** [comment]
hey bobdenotter.i am using the default base-2016 theme that is provided with the .zip package.facing issues with it .it gets redirected to a wrong url . and yes as u mentioned the reference   i do see the content like this person has mentioned #5924   .wrong canonical url with a subfolder url .

**Comment by trollix** [comment]
hey bobdenotter,
the bug is  with  the default base-2016 theme.
Thank you for the good work

**Comment by bobdenotter** [comment]
I've taken the liberty of updating the title from "Frontend not loading in Bolt 3.2.2" to a more descriptive one. :-)

**Comment by prbt2016** [comment]
Hey thanks bobdenotter .that's cool !

**Comment by bobdenotter** [comment]
Pinging @CarsonF You know most about the `{{ assets }}`, right? I've been getting more reports of people being bitten by this.

**Comment by rompcik** [comment]
I don't think it's the asset itself, since it also affects url({{ headerimage }}).
Here's the path part of dump():
```  "paths" => array:41 [▼
    "root" => "/~irata/bolt-3.2.3/"
    "rootpath" => "/home/irata/public_html/bolt-3.2.3"
    "apppath" => "/home/irata/public_html/bolt-3.2.3/vendor/bolt/bolt/app"
    "extensionsconfig" => "/home/irata/public_html/bolt-3.2.3/app/config/extensions"
    "extensionsconfigpath" => "/home/irata/public_html/bolt-3.2.3/app/config/extensions"
    "extensionspath" => "/home/irata/public_html/bolt-3.2.3/extensions"
    "filespath" => "/home/irata/public_html/bolt-3.2.3/files"
    "web" => "/home/irata/public_html/bolt-3.2.3"
    "webpath" => "/home/irata/public_html/bolt-3.2.3"
    "cache" => "/home/irata/public_html/bolt-3.2.3/app/cache"
    "cachepath" => "/home/irata/public_html/bolt-3.2.3/app/cache"
    "config" => "/home/irata/public_html/bolt-3.2.3/app/config"
    "configpath" => "/home/irata/public_html/bolt-3.2.3/app/config"
    "src" => "/home/irata/public_html/bolt-3.2.3/vendor/bolt/bolt/src"
    "srcpath" => "/home/irata/public_html/bolt-3.2.3/vendor/bolt/bolt/src"
    "database" => "/home/irata/public_html/bolt-3.2.3/app/database"
    "databasepath" => "/home/irata/public_html/bolt-3.2.3/app/database"
    "themebase" => "/home/irata/public_html/bolt-3.2.3/theme"
    "themebasepath" => "/home/irata/public_html/bolt-3.2.3/theme"
    "view" => "/bolt-public/view/"
    "viewpath" => "/home/irata/public_html/bolt-3.2.3/bolt-public/view"
    "composer" => "/home/irata/public_html/bolt-3.2.3/vendor/bolt/bolt"
    "composerpath" => "/home/irata/public_html/bolt-3.2.3/vendor/bolt/bolt"
    "app" => "/~irata/bolt-3.2.3/bolt-public/"
    "files" => "/~irata/bolt-3.2.3/files/"
    "themepath" => "/home/irata/public_html/bolt-3.2.3/theme/base-2016"
    "templatespath" => "/home/irata/public_html/bolt-3.2.3/theme/base-2016"
    "extensions" => "/~irata/bolt-3.2.3/extensions/"
    "async" => "/~irata/bolt-3.2.3/async/"
    "upload" => "/~irata/bolt-3.2.3/upload/"
    "bolt" => "/~irata/bolt-3.2.3/bolt/"
    "theme" => "/~irata/bolt-3.2.3/theme/base-2016/"
    "themes" => "/theme/"
    "current" => "/~irata/bolt-3.2.3/"
    "currenturl" => "http://localhost/~irata/bolt-3.2.3/"
    "hosturl" => "http://localhost"
    "rooturl" => "http://localhost/~irata/bolt-3.2.3/"
    "canonicalurl" => "http://localhost/~irata/bolt-3.2.3/"
    "canonical" => "http://localhost"
    "protocol" => "http"
    "hostname" => "localhost"
  ]
```

**Comment by CarsonF** [comment]
@rompcik Can you post `{{ dump(app.request.basePath) }}`?

**Comment by rompcik** [comment]
```{{ dump(app.request.basePath) }}``` is "/~irata/bolt-3.2.3"

**Comment by CarsonF** [comment]
@rompcik @prbt2016 Is this from a fresh install of 3.2? How are you bootstrapping the app? What does `public/web.php` look like? Are you using `.bolt.yml` file?

**Comment by bobdenotter** [comment]
@CarsonF Do you just need more info, or are you having trouble to replicate it? In my case replicating is pretty easy, by just installing Bolt in a subfolder, like `http://localhost/foo/`.

**Comment by sbibauw** [comment]
I'm having the same issue with a new install of Bolt 3.2.3, downloaded as .zip and installed to a subfolder of my root apache folder. For what I've seen (I'm totally new to Bolt):

- It occurs only with `base-2016` theme, not with others (e.g., no problem with `mikescops/bootstrapbolttheme`).
- The problem seem to be occurring mostly with the CSS urls.
- The asset path is duplicated, hence the error. Rather than referring to `/site/bolt/public/theme/...` it refers to `/site/bolt/public/site/bolt/public/theme/...`!

**Comment by CarsonF** [comment]
Which _specific_ version of PHP are yall using? I cannot reproduce this on 5.5.38, 5.6.27, or 7.0.13

**Comment by sbibauw** [comment]
I tried with PHP 5.6.27 and 7.0.12, it is exactly the same. 
Just to be clear, I installed via the "easy way" (i.e., `curl`, `tar`, `php app/nut init`).

**Comment by CarsonF** [comment]
@sbibauw Can you try modifying `vendor/bolt/bolt/app/web.php` and add this snippet after line 16?

```php
    // Fix server variables for PHP built-in server so base path is correctly determined.
    $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
    $frame = end($trace);
    // Absolute path to entry file
    $_SERVER['SCRIPT_FILENAME'] = $frame['file'];
    // Relative path to entry file from document root (dir the server is point to)
    $_SERVER['SCRIPT_NAME'] = preg_replace("#^{$_SERVER['DOCUMENT_ROOT']}#", '', $_SERVER['SCRIPT_FILENAME']);
```

It should look like this afterwards https://github.com/bolt/bolt/blob/release/3.3/app/web.php

Does that fix it?

**Comment by sbibauw** [comment]
@CarsonF I added the snippet to `web.php` but the problem remains exactly the same. Sorry.

I don't know if it might help, but on the sample homepage, I have 6 ressources that cannot load: css files, js files and the images/01.jpg etc. from the background. This said, the one asset that _does_ load correctly is the `<img src="/site/bolt/public/thumbs/1200x500c/california-foggy-golden-gate-bridge-2771.jpg">` on one of the sample articles.

**Comment by CarsonF** [comment]
@sbibauw Can you dump your $_SERVER vars here?

**Comment by sbibauw** [comment]
@CarsonF Here it goes (I am using MAMP):

    [HTTP_HOST] => localhost:8888
    [HTTP_CONNECTION] => keep-alive
    [HTTP_CACHE_CONTROL] => max-age=0
    [HTTP_UPGRADE_INSECURE_REQUESTS] => 1
    [HTTP_USER_AGENT] => Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.143 Safari/537.36
    [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
    [HTTP_DNT] => 1
    [HTTP_ACCEPT_ENCODING] => gzip, deflate, sdch
    [HTTP_ACCEPT_LANGUAGE] => fr-FR,fr;q=0.8,en-GB;q=0.6,en;q=0.4,en-US;q=0.2,es;q=0.2,nl;q=0.2
    [PATH] => /usr/bin:/bin:/usr/sbin:/sbin
    [SERVER_SIGNATURE] => 
    [SERVER_SOFTWARE] => Apache/2.2.31 (Unix) mod_wsgi/3.5 Python/2.7.12 PHP/7.0.12 mod_ssl/2.2.31 OpenSSL/1.0.2j DAV/2 mod_fastcgi/2.4.6 mod_perl/2.0.9 Perl/v5.24.0
    [SERVER_NAME] => localhost
    [SERVER_ADDR] => ::1
    [SERVER_PORT] => 8888
    [REMOTE_ADDR] => ::1
    [DOCUMENT_ROOT] => ~/Dropbox/dev
    [SERVER_ADMIN] => you@example.com
    [SCRIPT_FILENAME] => ~/Dropbox/dev/amap/tests/bolt/servervars.php
    [REMOTE_PORT] => 61308
    [GATEWAY_INTERFACE] => CGI/1.1
    [SERVER_PROTOCOL] => HTTP/1.1
    [REQUEST_METHOD] => GET
    [QUERY_STRING] => 
    [REQUEST_URI] => /amap/tests/bolt/servervars.php
    [SCRIPT_NAME] => /amap/tests/bolt/servervars.php
    [PHP_SELF] => /amap/tests/bolt/servervars.php
    [REQUEST_TIME_FLOAT] => 1481756955.16
    [REQUEST_TIME] => 1481756955

**Comment by sbibauw** [comment]
Actually, I just did a new installation of Bolt (CLI-install) in my `DOCUMENT_ROOT` (`/Users/username/Dropbox/dev/`) and I get exactly the same problem. Because the whole Bolt folder is set at the root level, the `public` folder is a subfolder of root and gets duplicated in the assets' path (`public/public/theme/style.css`).

**Comment by CarsonF** [comment]
Ok I can reproduce that. Although I'm not sure it's a bug...That seems like a misconfiguration.

**Comment by bobdenotter** [comment]
@CarsonF For me it breaks on PHP 5.5, 5.6 and 7 as well. Are you sure you're using a _subfolder_ when trying to replicate this?

**Comment by rompcik** [comment]
Seems like Theme's url has been assembled incorrectly, I suspect function setThemePath () in vendor/bolt/bolt/src/Configuration/ResourceManager.php.
Changing Line 428 fixes it for me - but only for asset type 'theme' -, heres the diff:
```428c428
<         $themeUrl = isset($generalConfig['theme_path']) ? $generalConfig['theme_path'] : $this->getUrl('root') . 'theme';
---
>         $themeUrl = isset($generalConfig['theme_path']) ? $generalConfig['theme_path'] : 'theme';
```
With other assets types (files, bolt, extensions)  the error is still present
@CasonF: Installation is Debian 8 (Jessie) default with Apache 2.4.10 and PHP 5.6.27

**Comment by CarsonF** [comment]
@rompcik Is $app['resources']->urlPrefix empty for you? What is the value of `getUrl('root')`?

I'm struggling with this because I refactored all this in 3.3. It's a mess in 3.2.

**Comment by CarsonF** [comment]
Ok with @bobdenotter's help, I've figured it out. I have indeed fixed this in 3.3. I'll backport the fix to 3.2 later today.
